File: android/src/main/kotlin/com/netki/netki_sdk/OnBoardIdBridge.kt
package com.netki.netki_sdk

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.google.accompanist.permissions.ExperimentalPermissionsApi
import com.netki.OnBoardId
import com.netki.model.AdditionalDataField
import com.netki.model.Environment
import com.netki.model.ErrorType
import com.netki.model.IdType
import com.netki.model.ResultInfo
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@OptIn(ExperimentalPermissionsApi::class)
class OnBoardIdBridge : EventChannel.StreamHandler, PluginRegistry.ActivityResultListener {
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var pendingResult: Result? = null

    companion object {
        private const val IDENTIFICATION_REQUEST_CODE = 1001
        private const val BIOMETRICS_REQUEST_CODE = 1002
    }

    fun setContext(context: Context) {
        this.context = context
    }

    fun setActivityBinding(binding: ActivityPluginBinding) {
        activityBinding = binding
        binding.addActivityResultListener(this)
    }

    fun clearActivityBinding() {
        activityBinding?.removeActivityResultListener(this)
        activityBinding = null
    }

    fun handle(call: MethodCall, result: Result, activity: Activity?) {
        when (call.method) {
            "initialize" -> initialize(call, result)
            "configureWithToken" -> configureWithToken(call, result)
            "requestSecurityCode" -> requestSecurityCode(call, result)
            "bypassSecurityCode" -> bypassSecurityCode(call, result)
            "confirmSecurityCode" -> confirmSecurityCode(call, result)
            "setBusinessMetadata" -> setBusinessMetadata(call, result)
            "startIdentification" -> startIdentification(call, result, activity)
            "submitIdentification" -> submitIdentification(call, result)
            "startBiometrics" -> startBiometrics(call, result, activity)
            "submitBiometrics" -> submitBiometrics(call, result)
            "getAvailableIdTypes" -> getAvailableIdTypes(result)
            "getAvailableCountries" -> getAvailableCountries(result)
            "getBusinessConfiguration" -> getBusinessConfiguration(result)
            "setLocation" -> setLocation(call, result)
            "setLivenessSettings" -> setLivenessSettings(call, result)
            "setClientGuid" -> setClientGuid(call, result)
            else -> result.notImplemented()
        }
    }

    // MARK: - Initialization & Configuration

    private fun initialize(call: MethodCall, result: Result) {
        val ctx = context ?: run {
            result.error("NO_CONTEXT", "Context not available", null)
            return
        }

        val environmentStr = call.argument<String>("environment")
        val environment = environmentStr?.let { Environment.valueOf(it) }

        OnBoardId.initialize(ctx, environment)
        result.success(null)
    }

    private fun configureWithToken(call: MethodCall, result: Result) {
        val token = call.argument<String>("token") ?: run {
            result.error("INVALID_ARGUMENTS", "Token is required", null)
            return
        }
        val accessCode = call.argument<String>("accessCode") ?: ""

        CoroutineScope(Dispatchers.Main).launch {
            val resultInfo = OnBoardId.configureWithToken(token, accessCode)
            result.success(Mappers.resultInfoToMap(resultInfo))
        }
    }

    // MARK: - Security Code

    private fun requestSecurityCode(call: MethodCall, result: Result) {
        val phoneNumber = call.argument<String>("phoneNumber") ?: run {
            result.error("INVALID_ARGUMENTS", "Phone number is required", null)
            return
        }

        CoroutineScope(Dispatchers.Main).launch {
            val resultInfo = OnBoardId.requestSecurityCode(phoneNumber)
            result.success(Mappers.resultInfoToMap(resultInfo))
        }
    }

    private fun bypassSecurityCode(call: MethodCall, result: Result) {
        val phoneNumber = call.argument<String>("phoneNumber") ?: run {
            result.error("INVALID_ARGUMENTS", "Phone number is required", null)
            return
        }

        OnBoardId.bypassSecurityCode(phoneNumber)
        result.success(null)
    }

    private fun confirmSecurityCode(call: MethodCall, result: Result) {
        val phoneNumber = call.argument<String>("phoneNumber") ?: run {
            result.error("INVALID_ARGUMENTS", "Phone number is required", null)
            return
        }
        val securityCode = call.argument<String>("securityCode") ?: run {
            result.error("INVALID_ARGUMENTS", "Security code is required", null)
            return
        }

        CoroutineScope(Dispatchers.Main).launch {
            val resultInfo = OnBoardId.confirmSecurityCode(phoneNumber, securityCode)
            result.success(Mappers.resultInfoToMap(resultInfo))
        }
    }

    // MARK: - Business Metadata

    private fun setBusinessMetadata(call: MethodCall, result: Result) {
        val metadata = call.argument<Map<String, String>>("metadata") ?: run {
            result.error("INVALID_ARGUMENTS", "Metadata is required", null)
            return
        }

        OnBoardId.setBusinessMetadata(metadata)
        result.success(null)
    }

    // MARK: - Identification

    private fun startIdentification(call: MethodCall, result: Result, activity: Activity?) {
        if (activity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        val idTypeStr = call.argument<String>("idType") ?: run {
            result.error("INVALID_ARGUMENTS", "ID type is required", null)
            return
        }
        val idCountryMap = call.argument<Map<String, Any>>("idCountry") ?: run {
            result.error("INVALID_ARGUMENTS", "ID country is required", null)
            return
        }

        val idType = IdType.entries.firstOrNull { it.type == idTypeStr } ?: run {
            result.error("INVALID_ARGUMENTS", "Invalid ID type", null)
            return
        }
        val idCountry = Mappers.mapToIdCountry(idCountryMap)

        pendingResult = result
        val intent = OnBoardId.getIdentificationIntent(idType, idCountry)
        activity.startActivityForResult(intent, IDENTIFICATION_REQUEST_CODE)
    }

    private fun submitIdentification(call: MethodCall, result: Result) {
        val additionalDataMap = call.argument<Map<String, String>>("additionalData") ?: emptyMap()
        val additionalData = Mappers.mapToAdditionalDataFields(additionalDataMap)

        CoroutineScope(Dispatchers.Main).launch {
            val resultInfo = OnBoardId.submitIdentification(additionalData)
            result.success(Mappers.resultInfoToMap(resultInfo))
        }
    }

    // MARK: - Biometrics

    private fun startBiometrics(call: MethodCall, result: Result, activity: Activity?) {
        if (activity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        val transactionId = call.argument<String>("transactionId") ?: run {
            result.error("INVALID_ARGUMENTS", "Transaction ID is required", null)
            return
        }

        pendingResult = result
        val intent = OnBoardId.getBiometricsIntent(transactionId)
        activity.startActivityForResult(intent, BIOMETRICS_REQUEST_CODE)
    }

    private fun submitBiometrics(call: MethodCall, result: Result) {
        val additionalDataMap = call.argument<Map<String, String>>("additionalData") ?: emptyMap()
        val additionalData = Mappers.mapToAdditionalDataFields(additionalDataMap)

        CoroutineScope(Dispatchers.Main).launch {
            val resultInfo = OnBoardId.submitBiometrics(additionalData)
            result.success(Mappers.resultInfoToMap(resultInfo))
        }
    }

    // MARK: - Data Retrieval

    private fun getAvailableIdTypes(result: Result) {
        val idTypes = OnBoardId.getAvailableIdTypes()
        val idTypeStrings = idTypes.map { it.type }
        result.success(idTypeStrings)
    }

    private fun getAvailableCountries(result: Result) {
        val countries = OnBoardId.getAvailableCountries()
        val countryMaps = countries.map { Mappers.idCountryToMap(it) }
        result.success(countryMaps)
    }

    private fun getBusinessConfiguration(result: Result) {
        val config = OnBoardId.getBusinessConfiguration()
        result.success(Mappers.businessConfigurationToMap(config))
    }

    // MARK: - Settings

    private fun setLocation(call: MethodCall, result: Result) {
        val lat = call.argument<String>("lat") ?: run {
            result.error("INVALID_ARGUMENTS", "Latitude is required", null)
            return
        }
        val lon = call.argument<String>("lon") ?: run {
            result.error("INVALID_ARGUMENTS", "Longitude is required", null)
            return
        }

        OnBoardId.setLocation(lat, lon)
        result.success(null)
    }

    private fun setLivenessSettings(call: MethodCall, result: Result) {
        val enabled = call.argument<Boolean>("enabled") ?: run {
            result.error("INVALID_ARGUMENTS", "Enabled flag is required", null)
            return
        }

        OnBoardId.setLivenessSettings(enabled)
        result.success(null)
    }

    private fun setClientGuid(call: MethodCall, result: Result) {
        val clientGuid = call.argument<String>("clientGuid") ?: run {
            result.error("INVALID_ARGUMENTS", "Client GUID is required", null)
            return
        }

        OnBoardId.setClientGuid(clientGuid)
        result.success(null)
    }

    // MARK: - Activity Result Handling

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode != IDENTIFICATION_REQUEST_CODE && requestCode != BIOMETRICS_REQUEST_CODE) {
            return false
        }

        when (resultCode) {
            Activity.RESULT_OK -> {
                @Suppress("UNCHECKED_CAST", "DEPRECATION")
                val pictures = data?.getSerializableExtra(ResultInfo.ExtraData.PICTURES.description) as? List<com.netki.model.Picture>
                val extraData = mutableMapOf<String, Any>()
                pictures?.let { extraData["pictures"] = Mappers.picturesToList(it) }

                sendEvent(mapOf(
                    "event" to "onCaptureIdentificationSuccessfully",
                    "extraData" to extraData
                ))
            }
            Activity.RESULT_CANCELED -> {
                val errorType = data?.getSerializableExtra(ResultInfo.ExtraData.ERROR_TYPE.description) as? ErrorType
                val message = data?.getStringExtra(ResultInfo.ExtraData.MESSAGE.description)

                val resultInfo = ResultInfo.error(errorType, message)
                sendEvent(mapOf(
                    "event" to "onCaptureIdentificationCancelled",
                    "resultInfo" to Mappers.resultInfoToMap(resultInfo)
                ))
            }
        }

        pendingResult?.success(null)
        pendingResult = null
        return true
    }

    // MARK: - Event Sending

    private fun sendEvent(event: Map<String, Any?>) {
        eventSink?.success(event)
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: ios/Classes/OnBoardIdBridge.swift
import Flutter
import UIKit
import NetkiSDK
import SwiftUI

class OnBoardIdBridge: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?

    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            initialize(call: call, result: result)
        case "configureWithToken":
            configureWithToken(call: call, result: result)
        case "requestSecurityCode":
            requestSecurityCode(call: call, result: result)
        case "bypassSecurityCode":
            bypassSecurityCode(call: call, result: result)
        case "confirmSecurityCode":
            confirmSecurityCode(call: call, result: result)
        case "setBusinessMetadata":
            setBusinessMetadata(call: call, result: result)
        case "startIdentification":
            startIdentification(call: call, result: result)
        case "submitIdentification":
            submitIdentification(call: call, result: result)
        case "startBiometrics":
            startBiometrics(call: call, result: result)
        case "submitBiometrics":
            submitBiometrics(call: call, result: result)
        case "getAvailableIdTypes":
            getAvailableIdTypes(result: result)
        case "getAvailableCountries":
            getAvailableCountries(result: result)
        case "getBusinessConfiguration":
            getBusinessConfiguration(result: result)
        case "setLocation":
            setLocation(call: call, result: result)
        case "setLivenessSettings":
            setLivenessSettings(call: call, result: result)
        case "setClientGuid":
            setClientGuid(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Initialization & Configuration

    private func initialize(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let environmentStr = args?["environment"] as? String

        var environment: Environments? = nil
        if let envStr = environmentStr {
            environment = Environments(rawValue: envStr)
        }

        OnBoardId.shared.initialize(environment: environment)
        result(nil)
    }

    private func configureWithToken(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let token = args["token"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Token is required", details: nil))
            return
        }

        let accessCode = args["accessCode"] as? String ?? ""

        Task {
            let resultInfo = await OnBoardId.shared.configureWithToken(token: token, accessCode: accessCode)
            DispatchQueue.main.async {
                result(Mappers.resultInfoToMap(resultInfo))
            }
        }
    }

    // MARK: - Security Code

    private func requestSecurityCode(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let phoneNumber = args["phoneNumber"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Phone number is required", details: nil))
            return
        }

        Task {
            let resultInfo = await OnBoardId.shared.requestSecurityCode(phoneNumber: phoneNumber)
            DispatchQueue.main.async {
                result(Mappers.resultInfoToMap(resultInfo))
            }
        }
    }

    private func bypassSecurityCode(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let phoneNumber = args["phoneNumber"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Phone number is required", details: nil))
            return
        }

        OnBoardId.shared.bypassSecurityCode(phoneNumber: phoneNumber)
        result(nil)
    }

    private func confirmSecurityCode(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let phoneNumber = args["phoneNumber"] as? String,
              let securityCode = args["securityCode"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Phone number and security code are required", details: nil))
            return
        }

        Task {
            let resultInfo = await OnBoardId.shared.confirmSecurityCode(
                phoneNumber: phoneNumber,
                securityCode: securityCode
            )
            DispatchQueue.main.async {
                result(Mappers.resultInfoToMap(resultInfo))
            }
        }
    }

    // MARK: - Business Metadata

    private func setBusinessMetadata(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let metadata = args["metadata"] as? [String: String] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Metadata is required", details: nil))
            return
        }

        OnBoardId.shared.setBusinessMetadata(businessMetadata: metadata)
        result(nil)
    }

    // MARK: - Identification

    private func startIdentification(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let idTypeStr = args["idType"] as? String,
              let idCountryMap = args["idCountry"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "ID type and country are required", details: nil))
            return
        }

        guard let idType = IdType(rawValue: idTypeStr) else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid ID type", details: nil))
            return
        }

        let idCountry = Mappers.mapToIdCountry(idCountryMap)

        let identificationView = OnBoardId.shared.getIdentificationView(
            idType: idType,
            idCountry: idCountry,
            identificationDelegate: self
        )

        presentView(identificationView, result: result)
    }

    private func submitIdentification(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let additionalDataMap = args?["additionalData"] as? [String: String] ?? [:]
        let additionalData = Mappers.mapToAdditionalDataFields(additionalDataMap)

        Task {
            let resultInfo = await OnBoardId.shared.submitIdentification(additionalData: additionalData)
            DispatchQueue.main.async {
                result(Mappers.resultInfoToMap(resultInfo))
            }
        }
    }

    // MARK: - Biometrics

    private func startBiometrics(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let transactionId = args["transactionId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Transaction ID is required", details: nil))
            return
        }

        let biometricsView = OnBoardId.shared.getBiometricsView(
            transactionId: transactionId,
            identificationDelegate: self
        )

        presentView(biometricsView, result: result)
    }

    private func submitBiometrics(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let additionalDataMap = args?["additionalData"] as? [String: String] ?? [:]
        let additionalData = Mappers.mapToAdditionalDataFields(additionalDataMap)

        Task {
            let resultInfo = await OnBoardId.shared.submitBiometrics(additionalData: additionalData)
            DispatchQueue.main.async {
                result(Mappers.resultInfoToMap(resultInfo))
            }
        }
    }

    // MARK: - Data Retrieval

    private func getAvailableIdTypes(result: @escaping FlutterResult) {
        let idTypes = OnBoardId.shared.getAvailableIdTypes()
        let idTypeStrings = idTypes.map { $0.rawValue }
        result(idTypeStrings)
    }

    private func getAvailableCountries(result: @escaping FlutterResult) {
        let countries = OnBoardId.shared.getAvailableCountries()
        let countryMaps = countries.map { Mappers.idCountryToMap($0) }
        result(countryMaps)
    }

    private func getBusinessConfiguration(result: @escaping FlutterResult) {
        let config = OnBoardId.shared.getBusinessConfiguration()
        result(Mappers.businessConfigurationToMap(config))
    }

    // MARK: - Settings

    private func setLocation(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let lat = args["lat"] as? String,
              let lon = args["lon"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Latitude and longitude are required", details: nil))
            return
        }

        OnBoardId.shared.setLocation(lat: lat, lon: lon)
        result(nil)
    }

    private func setLivenessSettings(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let enabled = args["enabled"] as? Bool else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Enabled flag is required", details: nil))
            return
        }

        OnBoardId.shared.setLivenessSettings(enabled: enabled)
        result(nil)
    }

    private func setClientGuid(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let clientGuid = args["clientGuid"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Client GUID is required", details: nil))
            return
        }

        OnBoardId.shared.setClientGuid(clientGuid: clientGuid)
        result(nil)
    }

    // MARK: - View Presentation

    private func presentView<V: View>(_ view: V, result: @escaping FlutterResult) {
        guard let rootViewController = UIApplication.shared.keyWindow?.rootViewController else {
            result(FlutterError(code: "NO_VIEW_CONTROLLER", message: "Cannot get root view controller", details: nil))
            return
        }

        let hostingController = UIHostingController(rootView: view)
        hostingController.modalPresentationStyle = .fullScreen
        rootViewController.present(hostingController, animated: true) {
            result(nil)
        }
    }

    private func dismissPresentedViewController() {
        DispatchQueue.main.async {
            if let rootViewController = UIApplication.shared.keyWindow?.rootViewController,
               let presentedVC = rootViewController.presentedViewController {
                presentedVC.dismiss(animated: true)
            }
        }
    }

    // MARK: - Event Sending

    private func sendEvent(_ event: [String: Any]) {
        DispatchQueue.main.async {
            self.eventSink?(event)
        }
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

// MARK: - IdentificationDelegate

extension OnBoardIdBridge: IdentificationDelegate {
    func onCaptureIdentificationSuccessfully(extraData: [String: Any]?) {
        dismissPresentedViewController()
        sendEvent([
            "event": "onCaptureIdentificationSuccessfully",
            "extraData": Mappers.sanitizeForFlutter(extraData) ?? [:]
        ])
    }

    func onCaptureIdentificationCancelled(resultInfo: ResultInfo) {
        dismissPresentedViewController()
        sendEvent([
            "event": "onCaptureIdentificationCancelled",
            "resultInfo": Mappers.resultInfoToMap(resultInfo)
        ])
    }
}
